compiler/optimizer/LocalOpts.cpp

```diff
@@ -7175,20 +7096,65 @@ void TR_InvariantArgumentPreexistence::processNode(TR::Node *node, TR::TreeTop *
       processIndirectLoad(node, treeTop, visitCount);
    else if (node->getOpCode().isCallIndirect())
       processIndirectCall(node, treeTop, visitCount);
+   }
+
+TR_YesNoMaybe TR_InvariantArgumentPreexistence::classIsCompatibleWithMethod(TR_OpaqueClassBlock* thisClazz, TR_ResolvedMethod* method)
+   {
+   TR_OpaqueClassBlock *clazz = method->containingClass();
+
+   if (clazz)
+      {
+      if (thisClazz)
+         {
+         return fe()->isInstanceOf(thisClazz, clazz, true);
+         }
+      }
+   return TR_maybe;
+   }
 
+bool TR_InvariantArgumentPreexistence::devirtualizeCall(TR::Node *node, TR::TreeTop *treeTop, TR_OpaqueClassBlock* clazz)
+   {
+   TR::MethodSymbol   *methodSymbol   = node->getSymbol()->castToMethodSymbol();
+   TR_ResolvedMethod *resolvedMethod = methodSymbol->getResolvedMethodSymbol()? methodSymbol->getResolvedMethodSymbol()->getResolvedMethod() : NULL;
+   TR_ASSERT(classIsCompatibleWithMethod(clazz, resolvedMethod) == TR_yes, "Class should be compatible with method");
+   TR::SymbolReference *symRef = node->getSymbolReference();
+   int32_t offset = symRef->getOffset();
+   TR_ResolvedMethod *refinedMethod = symRef->getOwningMethod(comp())->getResolvedVirtualMethod(comp(), clazz, offset);
+
+   if (!performTransformation(comp(), "%sspecialize and devirtualize invoke [%p] on currently fixed or final parameter\n", optDetailString(), node))
+      return false;
+
+   if (!refinedMethod->isSameMethod(resolvedMethod))
+      {
+      TR::SymbolReference *newSymRef =
+          getSymRefTab()->findOrCreateMethodSymbol
+          (symRef->getOwningMethodIndex(), -1, refinedMethod, TR::MethodSymbol::Virtual);
+      newSymRef->copyAliasSets(symRef, getSymRefTab());
+      newSymRef->setOffset(offset);
+      node->setSymbolReference(newSymRef);
+      node->devirtualizeCall(treeTop);
```

Comment:
Are we sure we won't devirtualize based on currently final (eg no overrides loaded yet) without adding a guard or some other logic to recover when this no longer holds?

Reply:
This is handled during CHTable commit in [openj9 CHTable.cpp](https://github.com/eclipse/openj9/blob/master/runtime/compiler/env/CHTable.cpp#L314)